langchain-tool-retrieval.ipynb

```diff
@@ -22,329 +45,415 @@
+      "\n",
+      "Create embeddings for each of these OpenBB tool descriptions...\n"
      ]
     }
    ],
    "source": [
-    "# Let's set-up our tool retrieval\n",
-    "fundamental_openbb_tools = map_openbb_collection_to_langchain_tools(\n",
-    "    \"/equity/fundamental\"\n",
-    ")\n",
-    "\n",
-    "peer_comparison_openbb_tools = map_openbb_collection_to_langchain_tools(\n",
-    "    '/equity/compare'\n",
-    ")\n",
-    "\n",
-    "estimate_openbb_tools = map_openbb_collection_to_langchain_tools(\n",
-    "    \"/equity/estimates\"\n",
+    "# can also give a single string, but since we want our agent to have more context we provide access to more functions\n",
+    "# TODO: We can create a \"universal_openbb_tools\" that allows access to all OpenBB functions since we use embeddings\n",
+    "# for the description of each, the context size is not an issue. However, with a bigger vector store, the accuracy may\n",
+    "# take a hit.\n",
+    "openbb_tools = map_openbb_collection_to_langchain_tools(\n",
+    "    openbb_commands_root = [\n",
+    "        \"/equity/fundamental\",\n",
+    "        \"/equity/compare\",\n",
+    "        \"/equity/estimates\"\n",
+    "    ]\n",
     ")\n",
+    "print(f\"A total of {len(openbb_tools)} OpenBB tools has been prepared for function calling\\n\")\n",
     "\n",
-    "openbb_tools = fundamental_openbb_tools + peer_comparison_openbb_tools + estimate_openbb_tools\n",
     "\n",
+    "print(\"Processing each OpenBB tool description into a list of docs...\\n\")\n",
+    "# Parse the description (i.e. docstring + output fields) for each of these tools\n",
     "docs = [\n",
     "    Document(page_content=t.description, metadata={\"index\": i})\n",
     "    for i, t in enumerate(openbb_tools)\n",
     "]\n",
-    "print(len(docs))\n",
     "\n",
-    "# Emperically, these Sentence embeddings can occasionally be better.\n",
-    "# (But it is a substantially heavier dependency than relying on OpenAI's API.)\n",
+    "print(\"Create embeddings for each of these OpenBB tool descriptions...\")\n",
+    "# Create embeddings from each of these function descriptions\n",
+    "# this will be important for when we want the agent to know what\n",
+    "# function to use for a particular query\n",
+    "vector_store = FAISS.from_documents(docs, OpenAIEmbeddings())\n",
+    "\n",
+    "# Note: Empirically, sentence embeddings can occasionally be better. But due to prompting\n",
+    "# requesting keywords, it has been found that OpenAIEmbeddings leads to better results.\n",
+    "# Plus using OpenAI's API is far lighter than relying on Hugging Face's SentenceTransformerEmbeddings\n",
     "# sentence_transformer = SentenceTransformerEmbeddings(model_name=\"all-MiniLM-L6-v2\")\n",
-    "# vector_store = FAISS.from_documents(docs, sentence_transformer)\n",
-    "vector_store = FAISS.from_documents(docs, OpenAIEmbeddings())"
+    "# vector_store = FAISS.from_documents(docs, sentence_transformer)\n"
```

Comment:
I suggest we remove this for now, since it doesn't add much (besides potential confusion). It can always be a discussion point. But I think right now the notebook should be laser-focused.